### DISCUSSION TITLE
[Best Practices] What is the best approach to handle complex business rules?

### DISCUSSION THREAD

**Discussion by gilsouza** [pmndrs/zustand#1299] [Q&A] [answered]
Hello!

I've some doubts about business rules and complexity that can or can't be added to the store and its actions:

```js
const courseSlice: StateCreator<
  AppStore,
  [['zustand/devtools', never]],
  [],
  CourseStore
> = set => ({
  currentCourseId: null,
  isLoadingCourse: false,
  disciplines: [],
  subModules: [],
  couseType: 'postgrad',
  progress: 20,
  // complex business logic -> update some states, like couseType and progress information
  // handle load state and orchestrate data fetches
  loadCourse: () => [...],
  switchCourse: () => [...],
  [...]
});

export { courseSlice };
```

1. Which would be the best way to handle business logic? Is using custom hooks, as facade to abstract that logics or put all inside actions?

2. ```switchCourse: () => [...],``` Could it trigger loadCourse after change currentCourseId or react way with useEffect in custom hook is a better approach?

3. What is the recommendation to make it easier to maintain and tests?

4. Is it the same recommendations to keep the state more consistent and predictable?

I know Zustand is unopinionated and that the answers come with trade offs. But, I count on your support to open my mind and bring insight from these questions =)

**Comment by dai-shi** [comment, marked as answer]
Just my two cents. If I had a complex business logic, I'd extract the logic as a pure function (not tied to zustand) and and add tests to it.

**Reply by gilsouza** [reply]
Thanks for the quick reply @dai-shi! You are the best!

I don't know if I understood correctly. 

When you say "not linked to zustand", do you mean that the pure function call would not be present in the action?
Would do you just action that would update the state, in this case?

**Reply by dai-shi** [reply]
I mean an action function will call the pure function, and update state. I'm not sure if I understand the situation, though.

**Reply by gilsouza** [reply]
Thanks @dai-shi! That makes sense to me.
